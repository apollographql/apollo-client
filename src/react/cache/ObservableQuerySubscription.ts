import equal from '@wry/equality';
import { Observer } from 'zen-observable-ts';
import { canonicalStringify } from '../../cache';
import {
  ApolloQueryResult,
  ObservableQuery,
  OperationVariables,
} from '../../core';
import {
  Concast,
  hasDirectives,
  iterateObserversSafely,
} from '../../utilities';

type Listener<TData> = (result: ApolloQueryResult<TData>) => void;

interface Subscription {
  unsubscribe: () => void;
}

interface DataSource<TData> {
  result: ApolloQueryResult<TData>;
  subscribe(observer: Observer<ApolloQueryResult<TData>>): Subscription;
}

interface ConcastDataSourceOptions {
  deferred: boolean;
}

class ObservableQueryDataSource<TData> implements DataSource<TData> {
  private observableQuery: ObservableQuery<TData>;
  private _result: ApolloQueryResult<TData>;

  constructor(observableQuery: ObservableQuery<TData>) {
    this.observableQuery = observableQuery;
    this._result = observableQuery.getCurrentResult();
  }

  get result() {
    return this._result;
  }

  subscribe(observer: Observer<ApolloQueryResult<TData>>) {
    return this.observableQuery.subscribe({
      next: () => {
        const result = this.observableQuery.getCurrentResult();

        if (!equal(this.result, result)) {
          this._result = result;
          observer.next?.(result);
        }
      },
    });
  }
}

class ConcastDataSource<TData> implements DataSource<TData> {
  private readonly concast: Concast<ApolloQueryResult<TData>>;
  private observers = new Set<Observer<ApolloQueryResult<TData>>>();
  private _result: ApolloQueryResult<TData>;
  private _completed = false;

  public readonly promise: Promise<ApolloQueryResult<TData>>;

  constructor(
    concast: Concast<ApolloQueryResult<TData>>,
    { deferred }: ConcastDataSourceOptions
  ) {
    this.concast = concast;
    this.promise = this.maybeWrapConcastWithCustomPromise(concast, {
      deferred,
    });

    this.subscribeToConcast();
  }

  get result() {
    return this._result;
  }

  get completed() {
    return this._completed;
  }

  subscribe(observer: Observer<ApolloQueryResult<TData>>) {
    if (!this.observers.has(observer)) {
      this.observers.add(observer);
    }

    if (this.completed) {
      observer.next?.(this.result);
      observer.complete?.();
    }

    return {
      unsubscribe: () => {
        this.observers.delete(observer);
      },
    };
  }

  private subscribeToConcast() {
    const subscription = this.concast.subscribe({
      next: (result) => {
        if (!equal(this.result, result)) {
          this._result = result;
          iterateObserversSafely(this.observers, 'next', result);
        }
      },
      complete: () => {
        this._completed = true;
        subscription.unsubscribe();
        iterateObserversSafely(this.observers, 'complete');
      },
    });
  }

  private maybeWrapConcastWithCustomPromise(
    concast: Concast<ApolloQueryResult<TData>>,
    { deferred }: { deferred: boolean }
  ) {
    if (deferred) {
      return new Promise<ApolloQueryResult<TData>>((resolve, reject) => {
        // Unlike `concast.promise`, we want to resolve the promise on the initial
        // chunk of the deferred query. This allows the component to unsuspend
        // when we get the initial set of data, rather than waiting until all
        // chunks have been loaded.
        const subscription = concast.subscribe({
          next: (value) => {
            resolve(value);
            subscription.unsubscribe();
          },
          error: reject,
        });
      });
    }

    return concast.promise;
  }
}

export class ObservableQuerySubscription<TData = any> {
  public result: ApolloQueryResult<TData>;
  public promise: Promise<ApolloQueryResult<TData>>;
  public readonly observable: ObservableQuery<TData>;

  private dataSources = new Map<string, ConcastDataSource<TData>>();
  private observableQueryDataSource: ObservableQueryDataSource<TData>;
  private listeners = new Set<Listener<TData>>();
  private subscriptions = new Set<Subscription>();
  private deferred: boolean;
  private subscription: Subscription;

  constructor(observable: ObservableQuery<TData>) {
    this.observable = observable;
    this.observableQueryDataSource = new ObservableQueryDataSource(observable);
    this.deferred = hasDirectives(['defer'], observable.query);
    this.result = observable.getCurrentResult();
    this.fetch(observable.options.variables);
  }

  fetch<TVariables extends OperationVariables = OperationVariables>(
    variables: TVariables | undefined
  ) {
    const key = canonicalStringify(variables);

    if (!this.dataSources.has(key)) {
      const concast = this.observable.reobserveAsConcast({ variables });
      this.dataSources.set(
        key,
        new ConcastDataSource(concast, { deferred: this.deferred })
      );
    }
    const dataSource = this.dataSources.get(key)!;

    this.streamResultsFrom(dataSource);
    this.promise = dataSource.promise;

    return dataSource.promise;
  }

  subscribe(listener: Listener<TData>) {
    this.listeners.add(listener);

    return () => {
      this.listeners.delete(listener);
    };
  }

  streamResultsFrom(dataSource: DataSource<TData>) {
    this.subscription?.unsubscribe();

    this.setResult(dataSource.result);

    this.subscription = dataSource.subscribe({
      next: (result) => this.setResult(result),
      complete: () => {
        this.streamResultsFrom(this.observableQueryDataSource);
      },
    });
  }

  dispose() {
    this.listeners.clear();
    this.subscriptions.forEach((subscription) => subscription.unsubscribe());
  }

  setResult(result: ApolloQueryResult<TData>) {
    if (!equal(this.result, result)) {
      this.result = result;
      this.deliver(result);
    }
  }

  private deliver(result: ApolloQueryResult<TData>) {
    this.listeners.forEach((listener) => listener(result));
  }
}
