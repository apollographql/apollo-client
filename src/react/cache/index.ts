export type { SuspenseCacheOptions } from './SuspenseCache.js';
export { getSuspenseCache } from './getSuspenseCache.js';

import { SuspenseCache as RealSuspenseCache } from './SuspenseCache.js';

// TODO: remove export with release 3.8
// replace with
// export type { SuspenseCache } from './SuspenseCache.js';
/**
 * @deprecated 
 * It is no longer necessary to create a `SuspenseCache` instance and pass it into the `ApolloProvider`.
 * Please remove this code from your application.
 *  
 * This export will be removed with the final 3.8 release.
 */
export class SuspenseCache extends RealSuspenseCache {
  constructor(){
    super();
    // throwing an error here instead of using invariant - we do not want this error
    // message to be link-ified, but to directly show up as bold as possible
    throw new Error(
      'It is no longer necessary to create a `SuspenseCache` instance and pass it into the `ApolloProvider`.\n' + 
      'Please remove this code from your application. \n\n' + 
      'This export will be removed with the final 3.8 release.'
    );
  }
}