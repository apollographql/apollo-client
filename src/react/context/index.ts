export * from './ApolloConsumer';
export * from './ApolloContext';
export * from './ApolloProvider';
