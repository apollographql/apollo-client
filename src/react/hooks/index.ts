export * from './useApolloClient';
export * from './useLazyQuery';
export * from './useMutation';
export * from './useQuery';
export * from './useSubscription';
