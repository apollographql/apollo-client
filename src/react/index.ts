import { invariant } from "ts-invariant";
import { DEV } from "../utilities";
invariant("boolean" === typeof DEV, DEV);

export {
  ApolloProvider,
  ApolloConsumer,
  getApolloContext,
  resetApolloContext,
  ApolloContextValue
} from './context';

export * from './hooks';

export {
  DocumentType,
  IDocumentDefinition,
  operationName,
  parser
} from './parser';

export * from './types/types';
