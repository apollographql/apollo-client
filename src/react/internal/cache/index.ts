export type { SuspenseCacheOptions } from "./SuspenseCache.js";
export type { CacheKey, QueryKey } from "./types.js";
export { getSuspenseCache } from "./getSuspenseCache.js";
export type { QueryReference, QueryRefPromise } from "./QueryReference.js";
export {
  InternalQueryReference,
  unwrapQueryRef,
  wrapQueryRef,
  updateWrappedQueryRef,
  getWrappedPromise,
} from "./QueryReference.js";
