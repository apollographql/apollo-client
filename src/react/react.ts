let React: typeof import('react');

// Apollo Client can be used without React, which means we want to make sure
// `react` is only imported/required if actually needed. To help with this
// the `react` module is lazy loaded using `requireReactLazily` when used by
// Apollo Client's React integration layer.
export function requireReactLazily(): typeof import('react') {
  return React || (React = require('react'));
}
