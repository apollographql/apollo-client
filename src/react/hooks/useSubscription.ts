import { useContext, useState, useRef, useEffect } from 'react';
import { DocumentNode } from 'graphql';

import { SubscriptionHookOptions } from '../types/types';
import { SubscriptionData } from '../data';
import { OperationVariables } from '../../core';
import { getApolloContext } from '../context';

export function useSubscription<TData = any, TVariables = OperationVariables>(
  subscription: DocumentNode,
  options?: SubscriptionHookOptions<TData, TVariables>
) {
  const context = useContext(getApolloContext());
  const updatedOptions = options
    ? { ...options, subscription }
    : { subscription };
  const [result, setResult] = useState({
    loading: !updatedOptions.skip,
    error: undefined,
    data: undefined
  });

  const subscriptionDataRef = useRef<SubscriptionData<TData, TVariables>>();
  function getSubscriptionDataRef() {
    if (!subscriptionDataRef.current) {
      subscriptionDataRef.current = new SubscriptionData<TData, TVariables>({
        options: updatedOptions,
        context,
        setResult
      });
    }
    return subscriptionDataRef.current;
  }

  const subscriptionData = getSubscriptionDataRef();
  subscriptionData.setOptions(updatedOptions, true);
  subscriptionData.context = context;

  useEffect(() => subscriptionData.afterExecute());
  useEffect(() => subscriptionData.cleanup.bind(subscriptionData), []);

  return subscriptionData.execute(result);
}
