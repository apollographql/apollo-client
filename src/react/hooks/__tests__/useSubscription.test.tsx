import React from 'react';
import { render, cleanup, wait } from '@testing-library/react';
import gql from 'graphql-tag';

import { ApolloClient } from '../../../core';
import { InMemoryCache as Cache } from '../../../cache';
import { ApolloProvider } from '../../context';
import { MockSubscriptionLink } from '../../../testing';
import { useSubscription } from '../useSubscription';

describe('useSubscription Hook', () => {
  afterEach(cleanup);

  it('should handle a simple subscription properly', async () => {
    const subscription = gql`
      subscription {
        car {
          make
        }
      }
    `;

    const results = ['Audi', 'BMW', 'Mercedes', 'Hyundai'].map(make => ({
      result: { data: { car: { make } } }
    }));

    const link = new MockSubscriptionLink();
    const client = new ApolloClient({
      link,
      cache: new Cache({ addTypename: false })
    });

    let renderCount = 0;
    const Component = () => {
      const { loading, data, error } = useSubscription(subscription);
      switch (renderCount) {
        case 0:
          expect(loading).toBe(true);
          expect(error).toBeUndefined();
          expect(data).toBeUndefined();
          break;
        case 1:
          expect(loading).toBe(false);
          expect(data).toEqual(results[0].result.data);
          break;
        case 2:
          expect(loading).toBe(false);
          expect(data).toEqual(results[1].result.data);
          break;
        case 3:
          expect(loading).toBe(false);
          expect(data).toEqual(results[2].result.data);
          break;
        case 4:
          expect(loading).toBe(false);
          expect(data).toEqual(results[3].result.data);
          break;
        default:
      }
      setTimeout(() => {
        renderCount <= results.length &&
          link.simulateResult(results[renderCount - 1]);
      });
      renderCount += 1;
      return null;
    };

    render(
      <ApolloProvider client={client}>
        <Component />
      </ApolloProvider>
    );

    return wait(() => {
      expect(renderCount).toBe(5);
    });
  });

  it('should cleanup after the subscription component has been unmounted', async () => {
    const subscription = gql`
      subscription {
        car {
          make
        }
      }
    `;

    const results = [
      {
        result: { data: { car: { make: 'Pagani' } } }
      }
    ];

    const link = new MockSubscriptionLink();
    const client = new ApolloClient({
      link,
      cache: new Cache({ addTypename: false })
    });

    let renderCount = 0;
    let onSubscriptionDataCount = 0;
    let unmount: any;

    const Component = () => {
      const { loading, data, error } = useSubscription(subscription, {
        onSubscriptionData() {
          onSubscriptionDataCount += 1;
        }
      });
      switch (renderCount) {
        case 0:
          expect(loading).toBe(true);
          expect(error).toBeUndefined();
          expect(data).toBeUndefined();
          link.simulateResult(results[0]);
          break;
        case 1:
          expect(loading).toBe(false);
          expect(data).toEqual(results[0].result.data);

          setTimeout(() => {
            expect(onSubscriptionDataCount).toEqual(1);

            // After the component has been unmounted, the internal
            // ObservableQuery should be stopped, meaning it shouldn't
            // receive any new data (so the onSubscriptionDataCount should
            // stay at 1).
            unmount();
            link.simulateResult(results[0]);
          });
          break;
        default:
      }
      renderCount += 1;
      return null;
    };

    unmount = render(
      <ApolloProvider client={client}>
        <Component />
      </ApolloProvider>
    ).unmount;

    return wait(() => {
      expect(onSubscriptionDataCount).toEqual(1);
    });
  });

  it('should never execute a subscription with the skip option', async () => {
    const subscription = gql`
      subscription {
        car {
          make
        }
      }
    `;

    const link = new MockSubscriptionLink();
    const client = new ApolloClient({
      link,
      cache: new Cache({ addTypename: false })
    });

    let renderCount = 0;
    let onSubscriptionDataCount = 0;
    let unmount: any;

    const Component = () => {
      const { loading, data, error } = useSubscription(subscription, {
        skip: true,
        onSubscriptionData() {
          onSubscriptionDataCount += 1;
        }
      });
      switch (renderCount) {
        case 0:
          expect(loading).toBe(false);
          expect(error).toBeUndefined();
          expect(data).toBeUndefined();
          setTimeout(() => {
            unmount();
          });
          break;
        default:
      }
      renderCount += 1;
      return null;
    };

    unmount = render(
      <ApolloProvider client={client}>
        <Component />
      </ApolloProvider>
    ).unmount;

    return wait(() => {
      expect(onSubscriptionDataCount).toEqual(0);
      expect(renderCount).toEqual(1);
    });
  });

  it('should create a subscription after skip has changed from true to a falsy value', async () => {
    const subscription = gql`
      subscription {
        car {
          make
        }
      }
    `;

    const results = [
      {
        result: { data: { car: { make: 'Pagani' } } }
      },
      {
        result: { data: { car: { make: 'Scoop' } } }
      }
    ];

    const link = new MockSubscriptionLink();
    const client = new ApolloClient({
      link,
      cache: new Cache({ addTypename: false })
    });

    let renderCount = 0;
    let unmount: any;

    const Component = () => {
      const [, triggerRerender] = React.useState(0);
      const [skip, setSkip] = React.useState(true);
      const { loading, data, error } = useSubscription(subscription, {
        skip
      });
      switch (renderCount) {
        case 0:
          expect(loading).toBe(false);
          expect(error).toBeUndefined();
          expect(data).toBeUndefined();
          setSkip(false);
          break;
        case 1:
          expect(loading).toBe(true);
          expect(error).toBeUndefined();
          expect(data).toBeUndefined();
          link.simulateResult(results[0]);
          break;
        case 2:
          expect(loading).toBe(false);
          expect(data).toEqual(results[0].result.data);
          setSkip(true);
          break;
        case 3:
          expect(loading).toBe(false);
          expect(data).toBeUndefined();
          expect(error).toBeUndefined();
          // ensure state persists across rerenders
          triggerRerender(i => i + 1);
          break;
        case 4:
          expect(loading).toBe(false);
          expect(data).toBeUndefined();
          expect(error).toBeUndefined();
          setSkip(false);
          break;
        case 5:
          expect(loading).toBe(true);
          expect(error).toBeUndefined();
          expect(data).toBeUndefined();
          link.simulateResult(results[1]);
          break;
        case 6:
          expect(loading).toBe(false);
          expect(error).toBeUndefined();
          expect(data).toEqual(results[1].result.data);
          setTimeout(() => {
            unmount();
          });
          break;
        default:
      }
      renderCount += 1;
      return null;
    };

    unmount = render(
      <ApolloProvider client={client}>
        <Component />
      </ApolloProvider>
    ).unmount;

    return wait(() => {
      expect(renderCount).toEqual(7);
    });
  });
});
