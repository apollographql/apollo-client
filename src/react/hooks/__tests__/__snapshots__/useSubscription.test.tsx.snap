// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useSubscription Hook should handle immediate completions gracefully 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "Missing field 'car' while writing result {}",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`useSubscription Hook should handle immediate completions with multiple subscriptions gracefully 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "Missing field 'car' while writing result {}",
    ],
    Array [
      "Missing field 'car' while writing result {}",
    ],
    Array [
      "Missing field 'car' while writing result {}",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
